source/game_sa/CTagManager.cpp

```diff
@@ -0,0 +1,243 @@
+
+    auto pTag = CTagManager::FindTagDesc(pEntity);
+    assert(pTag); // Originally the function would access unitialized memory, by clearing EAX and dereferencing pointer to [EAX + 0x4] right after that
+
+    return pTag->m_nAlpha;
+}
+
+void CTagManager::ResetAlpha(CEntity* pEntity)
+{
+    if (!pEntity->m_pRwAtomic)
+        return;
+
+    auto pTagDesc = CTagManager::FindTagDesc(pEntity);
+    CTagManager::SetAlpha(pEntity->m_pRwAtomic, pTagDesc->m_nAlpha);
+}
+
+int64_t CTagManager::GetPercentageTagged()
+{
+    return static_cast<int64_t>(static_cast<double>(CTagManager::ms_numTagged) / static_cast<double>(CTagManager::ms_numTags) * 100.0F);
+}
+
+int64_t CTagManager::GetPercentageTaggedInArea(CRect* pArea)
+{
+    int iTotalTags = 0;
+    int iTagged = 0;
+    for (int32_t i = CTagManager::ms_numTags - 1; i >= 0; --i) {
+        auto& pTagDesc = CTagManager::ms_tagDesc[i];
+        auto vecPos = CVector2D(pTagDesc.m_pEntity->GetPosition());
+        if (pArea->IsPointInside(vecPos)) {
+            ++iTotalTags;
+            if (pTagDesc.m_nAlpha > CTagManager::ucAlphaTagged)
+                ++iTagged;
+        }
+    }
+
+    return static_cast<int64_t>(static_cast<double>(iTagged) / static_cast<double>(iTotalTags) * 100.0F);
+}
+
+void CTagManager::UpdateNumTagged()
+{
+    CTagManager::ms_numTagged = 0;
+    if (!CTagManager::ms_numTags)
+        return;
+
+    for (int32_t i = CTagManager::ms_numTags - 1; i >= 0; --i) {
+        auto& pTagDesc = CTagManager::ms_tagDesc[i];
+        if (pTagDesc.m_nAlpha > CTagManager::ucAlphaTagged)
+            ++CTagManager::ms_numTagged;
+    }
+}
+
+void CTagManager::SetAlphaInArea(CRect* pArea, unsigned char ucAlpha)
+{
+    for (int32_t i = CTagManager::ms_numTags - 1; i >= 0; --i) {
+        auto& pTagDesc = CTagManager::ms_tagDesc[i];
+        auto vecPos = CVector2D(pTagDesc.m_pEntity->GetPosition());
+        if (pArea->IsPointInside(vecPos) && pTagDesc.m_pEntity->m_pRwAtomic) {
+            CTagManager::SetAlpha(pTagDesc.m_pEntity->m_pRwAtomic, ucAlpha);
+            pTagDesc.m_nAlpha = ucAlpha;
+        }
+    }
+
+    CTagManager::UpdateNumTagged();
+}
+
+void CTagManager::SetAlpha(CEntity* pEntity, unsigned char ucAlpha)
+{
+    if (pEntity->m_pRwAtomic)
+        CVisibilityPlugins::SetUserValue(pEntity->m_pRwAtomic, ucAlpha);
+
+    auto pTag = CTagManager::FindTagDesc(pEntity);
+    auto bChangedState = false;
+    if (ucAlpha > CTagManager::ucAlphaTagged && pTag->m_nAlpha <= CTagManager::ucAlphaTagged)
+        bChangedState = true;
+
+    pTag->m_nAlpha = ucAlpha;
+    CTagManager::UpdateNumTagged();
+
+    if (bChangedState && !TheCamera.m_bWideScreenOn) {
+        if (CTagManager::ms_numTagged == CTagManager::ms_numTags)
+            CGarages::TriggerMessage("TAG_ALL", -1, 5000, -1);
+        else
+            CGarages::TriggerMessage("TAG_ONE", CTagManager::ms_numTagged, 5000, CTagManager::ms_numTags);
+    }
+}
+
+CEntity* CTagManager::GetNearestTag(CVector* vecPos)
```

Comment:
```suggestion
CEntity* CTagManager::GetNearestTag(CVector const& vecPos)
```
I made the same mistake where I ignored the prototype info from android IDB. It's kinda nice to stay close to the original source. I didn't check it for all functions, but if the prototype is not the same as android, then please change it.

Reply:
Fixed